app/code/Magento/Cms/Block/BlockByIdentifier.php

```diff
@@ -0,0 +1,170 @@
+declare(strict_types=1);
+
+namespace Magento\Cms\Block;
+
+use Magento\Cms\Api\Data\BlockInterface;
+use Magento\Cms\Api\GetBlockByIdentifierInterface;
+use Magento\Cms\Model\Block as BlockModel;
+use Magento\Cms\Model\Template\FilterProvider;
+use Magento\Framework\DataObject\IdentityInterface;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Framework\View\Element\AbstractBlock;
+use Magento\Framework\View\Element\Context;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * This class is replacement of \Magento\Cms\Block\Block, that accepts only `string` identifier of CMS Block
+ */
+class BlockByIdentifier extends AbstractBlock implements IdentityInterface
+{
+    public const CACHE_KEY_PREFIX = 'CMS_BLOCK';
+
+    /**
+     * @var GetBlockByIdentifierInterface
+     */
+    private $blockByIdentifier;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var FilterProvider
+     */
+    private $filterProvider;
+
+    /**
+     * @var BlockInterface
+     */
+    private $cmsBlock;
+
+    /**
+     * @param GetBlockByIdentifierInterface $blockByIdentifier
+     * @param StoreManagerInterface $storeManager
+     * @param FilterProvider $filterProvider
+     * @param Context $context
+     * @param array $data
+     */
+    public function __construct(
+        GetBlockByIdentifierInterface $blockByIdentifier,
+        StoreManagerInterface $storeManager,
+        FilterProvider $filterProvider,
+        Context $context,
+        array $data = []
+    ) {
+        parent::__construct($context, $data);
+        $this->blockByIdentifier = $blockByIdentifier;
+        $this->storeManager = $storeManager;
+        $this->filterProvider = $filterProvider;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function _toHtml(): string
+    {
+        try {
+            return $this->filterOutput(
+                $this->getCmsBlock()->getContent()
+            );
+        } catch (NoSuchEntityException $e) {
+            return '';
+        }
+    }
+
+    /**
+     * Returns the value of `identifier` injected in `<block>` definition
+     *
+     * @return string|null
+     */
+    private function getIdentifier(): ?string
+    {
+        return $this->getdata('identifier') ?: null;
+    }
+
+    /**
+     * Filters the Content
+     *
+     * @param string $content
+     * @return string
+     * @throws NoSuchEntityException
+     */
+    private function filterOutput(string $content): string
+    {
+        return $this->filterProvider->getBlockFilter()
+            ->setStoreId($this->getCurrentStoreId())
+            ->filter($content);
+    }
+
+    /**
+     * Loads the CMS block by `identifier` provided as an argument
+     *
+     * @return BlockInterface|BlockModel
+     * @throws NoSuchEntityException
+     */
+    private function getCmsBlock(): BlockInterface
+    {
+        if (!$this->getIdentifier()) {
+            throw new NoSuchEntityException(
```

Comment:
A `NoSuchEntityException` seems a bit unrelated to the actual cause of the exception, it wasn't that the entity was't found by the provided identifier, the identifier was simply not there.

Might there be a more suitable exception that could be thrown here?

Reply:
Nice catch, this should actually be `InvalidArgumentException`

Reply:
I see what you did there! 😜